### PR TITLE
docs(sdk): Add Emacs documentation and emit a typescript-language-server wrapper for (Emacs') LSP support

### PR DESCRIPTION
**Pull request by martinjlowm**
This PR adds SDK documentation for Emacs which utilizes the generated binary wrappers for VS Code and supplies a new wrapper for https://github.com/theia-ide/typescript-language-server, enabling LSP support within Emacs.

Additionally, this also sorts SDK wrappers by name.

### CODE REVIEW

File: packages/gatsby/content/advanced/editor-sdks.md [hunk 1]
```diff
@@ -65,6 +65,49 @@ yarn pnpify --sdk
...
+                               `(:system ,(concat project-directory ".vscode/pnpify/typescript-language-server/lib/cli.js")))
+               (lsp-dependency 'typescript
+                               `(:system ,(concat project-directory ".vscode/pnpify/typescript/bin/tsserver")))
+
+               ;; Re-(start) LSP to pick up the dependency changes above.
+               (lsp)
+               )))))
+```
+
+4. You may have to install `vscode-jsonrpc` and `vscode-languageserver-protocol` manually if `typescript-language-server` fails to start (they are not listed in its dependencies as of this writing):
```
Comment (arcanis) [inline packages/gatsby/content/advanced/editor-sdks.md]: Would it make sense to add those dependencies to the [builtin `packageExtensions`](https://github.com/yarnpkg/berry/blob/master/packages/plugin-compat/sources/extensions.ts)?

Comment (martinjlowm) [reply packages/gatsby/content/advanced/editor-sdks.md]: Certainly! I'm pretty sure the experience becomes less confusing.

File: packages/plugin-compat/sources/extensions.ts [hunk 2]
```diff
@@ -52,4 +52,12 @@ export const packageExtensions: Array<[string, any]> = [
       [`@types/keyv`]: `^3.1.1`,
     },
   }],
+  // vscode-jsonrpc and vscode-languageserver-protocol are imported at runtime
+  // yet absent from typescript-language-server's dependencies
+  [`typescript-language-server@*`, {
+    dependencies: {
+      [`vscode-jsonrpc`]: `^5.0.1`,
```
Comment (arcanis) [inline packages/plugin-compat/sources/extensions.ts]: Are you sure for `vscode-jsonrpc`? From a quick look to their repo, it seems to only be used for type imports, so it shouldn't be imported at runtime 🤔

File: packages/plugin-compat/sources/extensions.ts [hunk 3]
```diff
@@ -52,4 +52,12 @@ export const packageExtensions: Array<[string, any]> = [
       [`@types/keyv`]: `^3.1.1`,
     },
   }],
+  // vscode-jsonrpc and vscode-languageserver-protocol are imported at runtime
+  // yet absent from typescript-language-server's dependencies
```
Comment (arcanis) [inline packages/plugin-compat/sources/extensions.ts]: Can I ask you to open a PR as well on `typescript-language-server` and reference it in the comment? I'd like to keep track of which builtin compat entries are fixed upstream 🙏

File: packages/plugin-compat/sources/extensions.ts (hunk 2 above)
Comment (martinjlowm) [reply packages/plugin-compat/sources/extensions.ts]: It's required from within `lib/lsp-protocol.calls.proposed.js`, the call stack is as follows.
```
- <project>/.yarn/cache/typescript-language-server-npm-0.4.0-c491a3c644-2.zip/node_modules/typescript-language-server/lib/lsp-protocol.calls.proposed.js
- <project>/.yarn/cache/typescript-language-server-npm-0.4.0-c491a3c644-2.zip/node_modules/typescript-language-server/lib/lsp-connection.js
- <project>/.yarn/cache/typescript-language-server-npm-0.4.0-c491a3c644-2.zip/node_modules/typescript-language-server/lib/cli.js
```

File: packages/plugin-compat/sources/extensions.ts (hunk 3 above)
Comment (martinjlowm) [reply packages/plugin-compat/sources/extensions.ts]: I actually considered that, but never got around to do it. But, absolutely, I'll do it right away :)

**Comment by arcanis** [comment]
Perfect, thanks @martinjlowm !

**Comment by gamb** [comment]
@martinjlowm As far as I can tell `typescript-language-server` doesn't seem to recognise PnP dependencies, is this a known issue? Perhaps one we could highlight in these docs. By comparison I find https://github.com/ananthakumaran/tide works a lot better with Yarn2+PnP.

**Comment by gamb** [comment]
** That is, dependencies aren't recognised even if I `yarn unplug` them

**Comment by gamb** [comment]
`*lsp-log*` shows the sdk language server is used:
```
The following clients were selected based on priority: (server-id ts-ls, priority -2)
Command "/Users/adamgamble/yarn-emacs-test/.yarn/sdks/typescript-language-server/lib/cli.js --tsserver-path /Users/adamgamble/yarn-emacs-test/.yarn/sdks/typescript/bin/tsserver --stdio" is present on the path.
Command "/Users/adamgamble/yarn-emacs-test/.yarn/sdks/typescript-language-server/lib/cli.js --tsserver-path /Users/adamgamble/yarn-emacs-test/.yarn/sdks/typescript/bin/tsserver --stdio" is present on the path.
Found the following clients for /Users/adamgamble/yarn-emacs-test/hooks/useApi.ts: (server-id ts-ls, priority -2)
The following clients were selected based on priority: (server-id ts-ls, priority -2)
Cancelling textDocument/codeAction(26) in hook post-command-hook
```

**Comment by martinjlowm** [comment]
I migrated to LSP because I found Tide to be very slow :(. I don't know if that is still the case.

Your setup looks correct, however, to see implementation details you would have to unplug specific dependencies. It's obviously not ideal and it would require a mapping of find-file to an implementation that supports the PnP interface. Do you know how/if Tide handles this properly?

Type resolution itself should work out of the box, since it is the tsserver entry that resolves those and it's already wrapped for PnP.

**Comment by gamb** [comment]
Yeah I'd rather LSP as it is faster but even with unplugged dependencies I can't seem to get LSP to find `react`… I've not got a good debug setup yet though, e.g. can't see LSP responses.

Tide is a little slow, although it is worth trying it with `(setq tide-server-max-response-length 4284928)` (or some big number) for larger projects - avoids some unnecessary timeouts due to response size.

Had some success using Tide to jump to definition inside zipped buffer… https://github.com/ananthakumaran/tide/issues/388

**Comment by gamb** [comment]
```diff
-    "typescript": "^3.8.3"
+    "typescript": "^3.8.3",
+    "typescript-language-server": "^0.4.0"
+  },
+  "dependenciesMeta": {
+    "react": {
+      "unplugged": true
+    },
+    "react-dom": {
+      "unplugged": true
+    }
```
Hmm LSP totally not picking up "react" still. `Cannot find module 'react'. [2307]`

**Comment by martinjlowm** [comment]
Sorry, I forgot about this - I have just returned from a 2 week vacation... have you had any luck getting this to work? It sounds to me as if LSP still uses an unpatched typescript "entry", which does support resolving virtual paths in zip archives.

```
/Users/adamgamble/yarn-emacs-test/.yarn/sdks/typescript/bin/tsserver --stdio
```
looks good to me though 🤔 

Unplugging is only necessary if you want to follow definitions.

**Comment by delucca** [comment]
@martinjlowm same issue here, check #2049

**Pull request merged**